call_option.go

```diff
@@ -1,118 +1,138 @@
-	s.Timeout = time.Duration(w)
-}
-
-// WithTimeout sets the client-side timeout for API calls if the call isn't
-// retrying.
-func WithTimeout(timeout time.Duration) CallOption {
-	return withTimeout(timeout)
+type boRetryer struct {
+	backoff Backoff
+	codes   []codes.Code
 }
 
-type withRetryCodes []codes.Code
-
-func (w withRetryCodes) Resolve(s *CallSettings) {
-	s.RetrySettings.RetryCodes = make(map[codes.Code]bool)
-	for _, code := range []codes.Code(w) {
-		s.RetrySettings.RetryCodes[code] = true
+func (r *boRetryer) Retry(err error) (time.Duration, bool) {
+	c := grpc.Code(err)
+	for _, rc := range r.codes {
+		if c == rc {
+			return r.backoff.Pause(), true
+		}
 	}
+	return 0, false
 }
 
-// WithRetryCodes sets a list of Google API canonical error codes upon which a
-// retry should be attempted. If nil, the call will not retry.
-func WithRetryCodes(retryCodes []codes.Code) CallOption {
-	return withRetryCodes(retryCodes)
-}
+// Backoff implements exponential backoff.
+// The wait time between retries is a random value between 0 and the "retry envelope".
+// The envelope starts at Initial and increases by the factor of Multiplier every retry,
+// but is capped at Max.
+type Backoff struct {
+	// Initial is the initial value of the retry envelope, defaults to 1 second.
+	Initial time.Duration
 
-type withDelayTimeoutSettings MultipliableDuration
+	// Max is the maximum value of the retry envelope, defaults to 30 seconds.
+	Max time.Duration
 
-func (w withDelayTimeoutSettings) Resolve(s *CallSettings) {
-	s.RetrySettings.BackoffSettings.DelayTimeoutSettings = MultipliableDuration(w)
-}
+	// Multiplier is the factor by which the retry envelope increases.
+	// It should be greater than 1 and defaults to 2.
+	Multiplier float64
 
-// WithDelayTimeoutSettings specifies:
-// - The initial delay time, in milliseconds, between the completion of
-//   the first failed request and the initiation of the first retrying
-//   request.
-// - The multiplier by which to increase the delay time between the
-//   completion of failed requests, and the initiation of the subsequent
-//   retrying request.
-// - The maximum delay time, in milliseconds, between requests. When this
-//   value is reached, `RetryDelayMultiplier` will no longer be used to
-//   increase delay time.
-func WithDelayTimeoutSettings(initial time.Duration, max time.Duration, multiplier float64) CallOption {
-	return withDelayTimeoutSettings(MultipliableDuration{initial, max, multiplier})
+	// cur is the current retry envelope
+	cur time.Duration
 }
 
-type withRPCTimeoutSettings MultipliableDuration
-
-func (w withRPCTimeoutSettings) Resolve(s *CallSettings) {
-	s.RetrySettings.BackoffSettings.RPCTimeoutSettings = MultipliableDuration(w)
+func (bo *Backoff) Pause() time.Duration {
+	if bo.Initial == 0 {
+		bo.Initial = time.Second
+	}
+	if bo.cur == 0 {
+		bo.cur = bo.Initial
+	}
+	if bo.Max == 0 {
+		bo.Max = 30 * time.Second
+	}
+	if bo.Multiplier == 0 {
```

Comment:
I think this makes sense. I don't think there's a case for setting multiplier < 1, exponential backoff is exponentially longer backoff, not shorter.